File: solutions/java/src/trafficsignalsystem/TrafficLight.java
package trafficsignalsystem;

public class TrafficLight {
    private SignalState state;
    private final Direction direction;

    public TrafficLight(Direction direction) {
        this.direction = direction;
        this.state = new RedState(); // Default initial state
    }

    public void setState(SignalState state) {
        this.state = state;
    }

    public SignalState getState() {
        return state;
    }

    public Direction getDirection() {
        return direction;
    }

    public void handle(TrafficSignalController controller) {
        state.handle(this, controller, direction);
    }
}

File: solutions/typescript/src/TrafficSignalSystem/TrafficLight.ts
import { DirectionEnum } from "./Direction";
import RedState from "./SignalState/RedState";
import SignalState from "./SignalState/SignalState";
import TrafficSignalController from "./TrafficSignalController";

export default class TrafficLight {
    private state: SignalState;
    private direction: DirectionEnum;

    constructor(direction: DirectionEnum) {
        this.direction = direction;
        this.state = new RedState(); // Default initial state
    }

    setState(state: SignalState) {
        this.state = state;
    }

    getState() {
        return this.state;
    }

    getDirection() {
        return this.direction;
    }

    handle(controller: TrafficSignalController) {
        this.state.handle(this, controller, this.direction);
    }
}
